StashRepository: Merge init() into the constructor

proski opened the pull request:
```
*  StashRepository: Merge init() into the constructor
   
   Jenkins documentation doesn't guarantee that start() would not be called
   after run() for a trigger.
   
   While it doesn't happen in freestyle projects, Jenkins would actually
   call start() after run() on the trigger in a pipeline project.
   
   That would create a new StashRepository object that would not have the
   StashApiClient object initialized. When the build completes, it would try
   to post the build result as a PR comment. That would throw an exception.
   
   To prevent it, remove the init() method. Move StashApiClient
   initialization inside the StashRepository constructor, so that the
   StashRepository object always has a valid StashApiClient object.
   
   Add WireMock as a test dependency. Check that the newly constructed
   StashRepository object would actually send an HTTP request.
```

That's one of the commits that make pipeline support possible. Extracted from #69 to help with review.

I realize that constructors are less testable than other methods, but the focus here is on fixing an issue with pipelines with minimal changes.

In the future, we can use factory pattern or generics to improve testability. Also, it might be possible not to replace the `StashRepository` object on the `start()` call for the same job. I'm not doing any optimization at this time, I'm making the code more robust.

jakub-bochenski commented on src/test/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepositoryTest.java:
```diff
@@ -437,4 +439,22 @@ public void startJobSkipsNullParameters() {
+    try {
+      newStashRepository.getTargetPullRequests();
+      fail();
+    } catch (RuntimeException e) {
```
This is a bit awkward way to ensure object is setup correctly :) 
Do you think we could somehow test this via getting a positive result? I guess one way would be to use a mock server, like I suggested for #93

proski replied:
We can use the factory pattern for StashApiClient. That would solve many issues with the unit tests. We won't need a special `StashRepository` constructor for unit tests. I really consider this unit test a placeholder for a proper unit test.

I can use wiremock in the meantime. Using wiremock would make it an integration test, but it would be less fragile than analyzing a chain of exceptions.

jakub-bochenski replied:
Wiremock was the server I had in mind when I wrote "use a mock server" :)

proski replied:
Done. Using WireMock now.

I made sure that the unit test fails without the changes to StashRepository.

proski committed:
StashRepository: Merge init() into the constructor

Jenkins documentation doesn't guarantee that start() would not be called
after run() for a trigger.

While it doesn't happen in freestyle projects, Jenkins would actually
call start() after run() on the trigger in a pipeline project.

That would create a new StashRepository object that would not have the
StashApiClient object initialized. When the build completes, it would try
to post the build result as a PR comment. That would throw an exception.

To prevent it, remove the init() method. Move StashApiClient
initialization inside the StashRepository constructor, so that the
StashRepository object always has a valid StashApiClient object.

Add WireMock as a test dependency. Check that the newly constructed
StashRepository object would actually send an HTTP request.

proski commented:
The requested changes have been made. The unit test is using WireMock now.

jakub-bochenski commented on src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java:
```diff
@@ -89,6 +78,14 @@ public void init() {
             trigger.isIgnoreSsl());
   }
 
+  // For unit tests only
```
not quite true, as it's also called from the other constructor

proski replied:
The only outside use is by unit tests. I'm not changing that part.

jakub-bochenski replied:
I know, it's minor. However I think "visible for (unit) tests" would be a better comment

proski replied:
OK, I'll change it. I'm actually thinking of creating StashApiClient in StashBuildTrigger, it which case the comment would not apply, but that would be a bigger change.

jakub-bochenski replied:
Thanks. Though just to clarify: I wanted to just point it out, I don't think it's important enough that it has to be changed

jakub-bochenski commented on src/main/java/stashpullrequestbuilder/stashpullrequestbuilder/StashRepository.java:
```diff
@@ -68,17 +68,6 @@
     this(job, trigger, null);
-  }
-
-  // For unit tests only
```
This is weird, I though we already have fmt-maven-plugin active
Why do we still get those formatting changes?

proski replied:
That was just a patch rendering. I'm changing this PR to make the difference easier to see.

proski commented:
Reworked the PR to keep the constructors short. As a side effect, the diff should be easier to read.

proski commented:
Improved the comment for the constructor used in unit tests.

proski commented:
Using Java 8 version of WireMock and dynamic HTTP port for testing.

proski commented:
Added a verification that WireMock received a GET request.
